README.md
# Minimal mod allowing incompatibility with the base mod

This mod is incompatible with the base mod and removes nearly all features from the game.
It is meant for mod development, gameplay is not really possible with only this mod enabled.

This mod provides the minimal prototypes needed to load the game without the base mod.
Modders are encouraged to build their own mod on top of this one or depend on this mod to aid development of mods incompatible with the base mod.

**To load the game with this mod enabled, you have to disable the base mod: Open mod-list.json, find "base" and set "enabled" to false.
If you do not do this, this mod will most likely not be loaded by the game.**


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.2.1
Date: ????
  Changes:
---------------------------------------------------------------------------------------------------
Version: 0.2.0
Date: 2020-12-14
  Features:
    - Updated to 1.1.6.
---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 2020-09-02
  Features:
    - Updated to 1.0.0.
    - Enforced "dummy-x" naming scheme for more prototypes.
---------------------------------------------------------------------------------------------------
Version: 0.0.3
Date: 2020-07-03
  Features:
    - Updated to 0.18.34.
---------------------------------------------------------------------------------------------------
Version: 0.0.2
Date: 2020-04-27
  Features:
    - Updated to 0.18.21.
---------------------------------------------------------------------------------------------------
Version: 0.0.1
Date: 2020-04-10
  Major Features:
    - Initial release.


items.lua
-- This file contains all items, including ones that are core prototypes.
--   Core prototypes are named prototypes that the game requires to always exist, for example a "copper-cable" item.

local properties = require("common-properties")

-- These functions are applied directly to the prototype table.
-- This allows to easily create many prototypes that inherit from the same base class/prototype, without copy pasting properties.
local function add_item_properties(prototype)
  prototype = properties.add_icon(prototype)
  prototype.stack_size = 1
  return prototype
end

local function add_selection_tool_properties(prototype)
  prototype = add_item_properties(prototype)
  prototype.selection_color = properties.color()
  prototype.alt_selection_color = properties.color()
  prototype.selection_mode = "nothing"
  prototype.alt_selection_mode = "nothing"
  prototype.selection_cursor_box_type = "entity"
  prototype.alt_selection_cursor_box_type = "entity"
  return prototype
end

-- item prototypes, one of each, alphabetical by type
data:extend({
  add_item_properties
  {
    type = "ammo",
    name = "dummy-ammo",
    ammo_type =
    {
      category = "dummy-ammo-category"
    }
  },
  add_item_properties
  {
    type = "armor",
    name = "dummy-armor",
    infinite = true
  },
  add_selection_tool_properties
  {
    type = "blueprint",
    name = "dummy-blueprint"
  },
  add_item_properties
  {
    type = "blueprint-book",
    name = "dummy-blueprint-book",
    inventory_size = 1
  },
  add_item_properties
  {
    type = "capsule",
    name = "dummy-capsule",
    capsule_action =
    {
      type = "equipment-remote",
      equipment = "dummy-active-defense-equipment"
    }
  },
  add_selection_tool_properties
  {
    type = "copy-paste-tool",
    name = "dummy-copy-paste-tool"
  },
  add_selection_tool_properties
  {
    type = "deconstruction-item",
    name = "dummy-deconstruction-item"
  },
  add_item_properties
  {
    type = "gun",
    name = "dummy-gun",
    attack_parameters = properties.attack_parameters()
  },
  add_item_properties
  {
    type = "item",
    name = "dummy-item"
  },
  add_item_properties
  {
    type = "item-with-entity-data",
    name = "dummy-item-with-entity-data"
  },
  add_item_properties
  {
    type = "item-with-inventory",
    name = "dummy-item-with-inventory",
    inventory_size = 1
  },
  add_item_properties
  {
    type = "item-with-label",
    name = "dummy-item-with-label"
  },
  add_item_properties
  {
    type = "item-with-tags",
    name = "dummy-item-with-tags"
  },
  add_item_properties
  {
    type = "mining-tool",
    name = "dummy-mining-tool", -- deprecated prototype
    infinite = true
  },
  add_item_properties
  {
    type = "module",
    name = "dummy-module",
    tier = 1,
    effect = {},
    category = "dummy-module-category"
  },
  add_item_properties
  {
    type = "rail-planner",
    name = "dummy-rail-planner",
    straight_rail = "dummy-straight-rail",
    curved_rail = "dummy-curved-rail"
  },
  add_item_properties
  {
    type = "repair-tool",
    name = "dummy-repair-tool",
    infinite = true,
    speed = 1
  },
  add_selection_tool_properties
  {
    type = "selection-tool",
    name = "dummy-selection-tool"
  },
  add_item_properties
  {
    type = "spidertron-remote",
    name = "dummy-spidertron-remote",
    icon_color_indicator_mask = properties.sprite_filename
  },
  add_item_properties
  {
    type = "tool",
    name = "dummy-tool",
    infinite = true
  },
  add_selection_tool_properties
  {
    type = "upgrade-item",
    name = "dummy-upgrade-item"
  }
})

-- core item prototypes that are not contained in the above data:extend
data:extend({
  add_item_properties
  {
    type = "item",
    name = "copper-cable"
  },
  add_item_properties
  {
    type = "item",
    name = "red-wire"
  },
  add_item_properties
  {
    type = "item",
    name = "green-wire"
  }
})


equipment.lua
-- This file contains all equipment.

local properties = require("common-properties")

-- This functions is applied directly to the prototype table.
-- This allows to easily create many prototypes that inherit from the same base class/prototype, without copy pasting properties.
local function add_equipment_properties(prototype)
  prototype.sprite = properties.sprite()
  prototype.shape = {
    type = "full",
    width = 1,
    height = 1
  }
  prototype.categories = {"dummy-equipment-category"}
  prototype.energy_source = properties.electric_energy_source()
  prototype.take_result = "dummy-item"
  return prototype
end

data:extend({
  add_equipment_properties
  {
    type = "active-defense-equipment",
    name = "dummy-active-defense-equipment",
    automatic = false,
    ability_icon = properties.sprite(),
    attack_parameters = properties.attack_parameters()
  },
  add_equipment_properties
  {
    type = "battery-equipment",
    name = "dummy-battery-equipment"
  },
  add_equipment_properties
  {
    type = "belt-immunity-equipment",
    name = "dummy-belt-immunity-equipment",
    energy_consumption = "1J"
  },
  add_equipment_properties
  {
    type = "energy-shield-equipment",
    name = "dummy-energy-shield-equipment",
    energy_per_shield = "1J",
    max_shield_value = 1
  },
  add_equipment_properties
  {
    type = "generator-equipment",
    name = "dummy-generator-equipment",
    power = "1J"
  },
  add_equipment_properties
  {
    type = "movement-bonus-equipment",
    name = "dummy-movement-bonus-equipment",
    energy_consumption = "1J",
    movement_bonus = 1
  },
  add_equipment_properties
  {
    type = "night-vision-equipment",
    name = "dummy-night-vision-equipment",
    energy_input = "1J",
    color_lookup = {{1, "identity"}}
  },
  add_equipment_properties
  {
    type = "roboport-equipment",
    name = "dummy-roboport-equipment",
    charge_approach_distance = 1,
    charging_energy = "1J",
    construction_radius = 1,
    recharging_animation = properties.animation(),
    spawn_and_station_height = 1
  },
  add_equipment_properties
  {
    type = "solar-panel-equipment",
    name = "dummy-solar-panel-equipment",
    power = "1J"
  }
})


properties.lua
-- Note: None of these properties are meant for use in a normal mod,
--   because they mostly produce empty images and have no function in-game beyond being loadable.
-- This file only contains properties that are large, or used very often.

-- All properties are based on the formats documented on the wiki: https://wiki.factorio.com/Prototype_overview

local properties = {}

properties.sound_filename = "__core__/sound/achievement-unlocked.ogg"
properties.sprite_filename = "__core__/graphics/empty.png"
properties.sprite_filename_32px = "__core__/graphics/factorio-icon.png"
properties.render_layer = "object"

-- This function is applied directly to the prototype table and adds the icon to it
properties.add_icon = function(prototype)
  prototype.icon = properties.sprite_filename
  prototype.icon_size = 1
  return prototype
end

-- Any property that is a table is returned via a function.
--   This means that everything gets a copy of the table, not a reference to the table.

properties.color = function() return {1, 1, 1, 1} end -- white
properties.bounding_box = function() return {{0, 0}, {0, 0}} end
-- splitter: bounding box width must be > 0.5, height must be > 1
properties.nonzero_bounding_box = function() return {{-1.0, -0.5}, {0.1, 0.1}} end
properties.vector = function() return {0, 0} end
properties.sound = function()
  return {filename = properties.sound_filename}
end
properties.sprite = function()
  return
  {
    filename = properties.sprite_filename,
    size = 1
  }
end
-- properties.sprite() is also used as SpriteVariations and Sprite4Way
properties.rotated_sprite = function()
  local sprite = properties.sprite()
  sprite.direction_count = 1
  return sprite
end
properties.rotated_sprite_custom_direction_count = function(direction_count)
  local sprite = properties.sprite()
  sprite.filename = properties.sprite_filename_32px
  sprite.direction_count = direction_count
  return sprite
end
properties.sprite_8_way = function()
  return
  {
    north = properties.sprite(),
    north_east = properties.sprite(),
    east = properties.sprite(),
    south_east = properties.sprite(),
    south = properties.sprite(),
    south_west = properties.sprite(),
    west = properties.sprite(),
    north_west = properties.sprite()
  }
end
properties.animation = function()
  return properties.sprite()
end
-- properties.animation() is also used as AnimationVariations and Animation4Way
properties.rotated_animation = function()
  local animation = properties.animation()
  animation.direction_count = 1
  return animation
end
properties.attack_parameters = function()
  return
  {
    type = "stream", -- required by fluid turret
    ammo_type = {category = "dummy-ammo-category"},
    range = 1,
    cooldown = 1,
    animation = properties.rotated_animation() -- required by unit
  }
end
properties.rail_piece_layers = function()
  return
  {
    metals = properties.sprite(),
    backplates = properties.sprite(),
    ties = properties.sprite(),
    stone_path = properties.sprite()
  }
end
properties.rail_pictures = function()
  return
  {
    straight_rail_horizontal = properties.rail_piece_layers(),
    straight_rail_vertical = properties.rail_piece_layers(),
    straight_rail_diagonal_left_top = properties.rail_piece_layers(),
    straight_rail_diagonal_right_top = properties.rail_piece_layers(),
    straight_rail_diagonal_right_bottom = properties.rail_piece_layers(),
    straight_rail_diagonal_left_bottom = properties.rail_piece_layers(),
    curved_rail_vertical_left_top = properties.rail_piece_layers(),
    curved_rail_vertical_right_top = properties.rail_piece_layers(),
    curved_rail_vertical_right_bottom = properties.rail_piece_layers(),
    curved_rail_vertical_left_bottom = properties.rail_piece_layers(),
    curved_rail_horizontal_left_top = properties.rail_piece_layers(),
    curved_rail_horizontal_right_top = properties.rail_piece_layers(),
    curved_rail_horizontal_right_bottom = properties.rail_piece_layers(),
    curved_rail_horizontal_left_bottom = properties.rail_piece_layers(),
    rail_endings = properties.sprite_8_way()
  }
end
properties.oriented_cliff_prototype = function()
  return
  {
    collision_bounding_box = properties.bounding_box(),
    pictures = {properties.sprite()},
    fill_volume = 1
  }
end
properties.electric_energy_source = function()
  return
  {
    type = "electric",
    buffer_capacity = "1J",
    input_flow_limit = "1W",
    -- this works even for entities that technically *output* energy
    usage_priority = "primary-input"
  }
end
properties.void_energy_source = function()
  return {type = "void"}
end
properties.burner_energy_source = function()
  return {type = "burner", fuel_inventory_size = 1}
end
properties.wire_connection_point = function()
  return
  {
    wire = {},
    shadow = {}
  }
end
properties.damage_prototype = function()
  return {amount = 1, type = "physical"}
end
properties.fluid_box = function()
  return {pipe_connections = {}}
end
properties.heat_buffer = function()
  return
  {
    -- because min temp defaults to 15 and this must be >= min temp
    max_temperature = 15,
    specific_heat = "1J",
    max_transfer = "1J"
  }
end

return properties

common-properties.lua
-- Note: None of these properties are meant for use in a normal mod,
--   because they mostly produce empty images and have no function in-game beyond being loadable.

-- All properties are based on the formats documented on the wiki: https://wiki.factorio.com/Prototype_overview

local properties = {}

properties.sound_filename = "__core__/sound/achievement-unlocked.ogg"
properties.sprite_filename = "__core__/graphics/empty.png"
properties.sprite_filename_32px = "__core__/graphics/icons/unknown.png"
properties.render_layer = "object"

-- This function is applied directly to the prototype table and adds the icon to it
properties.add_icon = function(prototype)
  prototype.icon = properties.sprite_filename
  prototype.icon_size = 1
  return prototype
end

-- Any property that is a table is returned via a function.
--   This means that everything gets a copy of the table, not a reference to the table.

properties.color = function() return {1, 1, 1, 1} end -- white
properties.bounding_box = function() return {{0, 0}, {0, 0}} end
-- splitter: bounding box width must be > 0.5, height must be > 1
properties.nonzero_bounding_box = function() return {{-0.95, -1.0}, {0.1, 0.1}} end
properties.vector = function() return {0, 0} end
properties.sound = function()
  return {filename = properties.sound_filename}
end
properties.sprite = function()
  return
  {
    filename = properties.sprite_filename,
    size = 1
  }
end
-- properties.sprite() is also used as SpriteVariations and Sprite4Way
properties.rotated_sprite = function()
  local sprite = properties.sprite()
  sprite.direction_count = 1
  return sprite
end
properties.rotated_sprite_custom_direction_count = function(direction_count)
  local sprite = properties.sprite()
  sprite.filename = properties.sprite_filename_32px
  sprite.direction_count = direction_count
  return sprite
end
properties.sprite_8_way = function()
  return
  {
    north = properties.sprite(),
    north_east = properties.sprite(),
    east = properties.sprite(),
    south_east = properties.sprite(),
    south = properties.sprite(),
    south_west = properties.sprite(),
    west = properties.sprite(),
    north_west = properties.sprite()
  }
end
properties.animation = function()
  return properties.sprite()
end
-- properties.animation() is also used as AnimationVariations and Animation4Way
properties.rotated_animation = function()
  local animation = properties.animation()
  animation.direction_count = 1
  return animation
end
properties.attack_parameters = function()
  return
  {
    type = "stream", -- required by fluid turret
    ammo_type = {category = "dummy-ammo-category"},
    range = 1,
    cooldown = 1,
    animation = properties.rotated_animation() -- required by unit
  }
end
properties.rail_piece_layers = function()
  return
  {
    metals = properties.sprite(),
    backplates = properties.sprite(),
    ties = properties.sprite(),
    stone_path = properties.sprite()
  }
end
properties.rail_pictures = function()
  return
  {
    straight_rail_horizontal = properties.rail_piece_layers(),
    straight_rail_vertical = properties.rail_piece_layers(),
    straight_rail_diagonal_left_top = properties.rail_piece_layers(),
    straight_rail_diagonal_right_top = properties.rail_piece_layers(),
    straight_rail_diagonal_right_bottom = properties.rail_piece_layers(),
    straight_rail_diagonal_left_bottom = properties.rail_piece_layers(),
    curved_rail_vertical_left_top = properties.rail_piece_layers(),
    curved_rail_vertical_right_top = properties.rail_piece_layers(),
    curved_rail_vertical_right_bottom = properties.rail_piece_layers(),
    curved_rail_vertical_left_bottom = properties.rail_piece_layers(),
    curved_rail_horizontal_left_top = properties.rail_piece_layers(),
    curved_rail_horizontal_right_top = properties.rail_piece_layers(),
    curved_rail_horizontal_right_bottom = properties.rail_piece_layers(),
    curved_rail_horizontal_left_bottom = properties.rail_piece_layers(),
    rail_endings = properties.sprite_8_way()
  }
end
properties.oriented_cliff_prototype = function()
  return
  {
    collision_bounding_box = properties.bounding_box(),
    pictures = {properties.sprite()},
    fill_volume = 1
  }
end
properties.electric_energy_source = function()
  return
  {
    type = "electric",
    buffer_capacity = "1J",
    input_flow_limit = "1W",
    -- this works even for entities that technically *output* energy
    usage_priority = "primary-input"
  }
end
properties.void_energy_source = function()
  return {type = "void"}
end
properties.burner_energy_source = function()
  return {type = "burner", fuel_inventory_size = 1}
end
properties.wire_connection_point = function()
  return
  {
    wire = {},
    shadow = {}
  }
end
properties.damage_prototype = function()
  return {amount = 1, type = "physical"}
end
properties.fluid_box = function()
  return {pipe_connections = {}}
end
properties.heat_buffer = function()
  return
  {
    -- because min temp defaults to 15 and this must be >= min temp
    max_temperature = 15,
    specific_heat = "1J",
    max_transfer = "1J"
  }
end

return properties

data.lua
local properties = require("common-properties")
require("items")
require("entities")
require("equipment")

-- The utility constants are defined in the core mod.
--   Removing this property means trigger-target-types "common" and "ground-unit" do not have to be created.
data.raw["utility-constants"]["default"].default_trigger_target_mask_by_type = nil
-- The "x-unknown" prototypes are defined in the core mod.
--   Setting the subgroup to an existing one means the default subgroup "fluid" does not have to be created.
data.raw["fluid"]["fluid-unknown"].subgroup = "other"
--   Setting the collision mask to something walkable means the game accepts this as the needed tile prototype.
data.raw["tile"]["tile-unknown"].collision_mask = {"ground-tile"}

-- Miscellaneous prototypes that are required by the required prototypes. Mostly categories.
data:extend({
  {
    type = "recipe-category",
    name = "crafting"
  },
  {
    type = "resource-category",
    name = "basic-solid"
  },
  {
    type = "fuel-category",
    name = "chemical"
  },
  {
    type = "ammo-category",
    name = "dummy-ammo-category"
  },
  {
    type = "equipment-category",
    name = "dummy-equipment-category"
  },
  {
    type = "module-category",
    name = "dummy-module-category"
  },
  {
    type = "optimized-particle",
    name = "dummy-optimized-particle",
    life_time = 2,
    pictures = properties.animation(),
    render_layer = properties.render_layer,
    render_layer_when_on_ground = properties.render_layer
  },
  properties.add_icon
  {
    type = "item-group",
    name = "dummy-item-group"
  }
})

-- The remaining core prototypes
--   Core prototypes are named prototypes that the game requires to always exist, for example a "signal-everything" virtual signal.
--   Core prototypes that are items or entities are included in items.lua and entities.lua respectively.
data:extend({
  {
    type = "item-subgroup",
    name = "other",
    group = "dummy-item-group"
  },
  {
    type = "equipment-grid",
    name = "dummy-equipment-grid",
    equipment_categories = {"dummy-equipment-category"},
    height = 1,
    width = 1
  },
  properties.add_icon
  {
    type = "virtual-signal",
    name = "signal-everything",
    subgroup = "other"
  },
  properties.add_icon
  {
    type = "virtual-signal",
    name = "signal-anything",
    subgroup = "other"
  },
  properties.add_icon
  {
    type = "virtual-signal",
    name = "signal-each",
    subgroup = "other"
  },
  {
    type = "damage-type",
    name = "physical"
  },
  {
    type = "damage-type",
    name = "impact"
  },
  {
    type = "trivial-smoke",
    name = "smoke-building",
    animation = properties.animation(),
    duration = 1,
    cyclic = true
  }
})

-- The map settings are required, despite not being a core prototype. These map settings are simply a copy from the base mod.
data:extend({
  {
    type="map-settings",
    name="map-settings",

    pollution=
    {
      enabled=true,
      diffusion_ratio=0.02,
      min_to_diffuse=15,
      ageing=1,
      expected_max_per_chunk=150,
      min_to_show_per_chunk=50,
      min_pollution_to_damage_trees = 60,
      pollution_with_max_forest_damage = 150,
      pollution_per_tree_damage = 50,
      pollution_restored_per_tree_damage = 10,
      max_pollution_to_restore_trees = 20,
      enemy_attack_pollution_consumption_modifier = 1
    },

    enemy_evolution=
    {
      enabled=true,
      time_factor = 0.000004,
      destroy_factor = 0.002,
      pollution_factor = 0.0000009
    },

    enemy_expansion=
    {
      enabled = true,
      max_expansion_distance = 7,
      friendly_base_influence_radius = 2,
      enemy_building_influence_radius = 2,
      building_coefficient = 0.1,
      other_base_coefficient = 2.0,
      neighbouring_chunk_coefficient = 0.5,
      neighbouring_base_chunk_coefficient = 0.4;
      max_colliding_tiles_coefficient = 0.9,
      settler_group_min_size = 5,
      settler_group_max_size = 20,
      min_expansion_cooldown = 4 * 3600,
      max_expansion_cooldown = 60 * 3600
    },

    unit_group=
    {
      min_group_gathering_time = 3600,
      max_group_gathering_time = 10 * 3600,
      max_wait_time_for_late_members = 2 * 3600,
      max_group_radius = 30.0,
      min_group_radius = 5.0,
      max_member_speedup_when_behind = 1.4,
      max_member_slowdown_when_ahead = 0.6,
      max_group_slowdown_factor = 0.3,
      max_group_member_fallback_factor = 3,
      member_disown_distance = 10,
      tick_tolerance_when_member_arrives = 60,
      max_gathering_unit_groups = 30,
      max_unit_group_size = 200
    },

    steering=
    {
      default=
      {
        radius = 1.2,
        separation_force = 0.005,
        separation_factor = 1.2,
        force_unit_fuzzy_goto_behavior = false
      },
      moving=
      {
        radius = 3,
        separation_force = 0.01,
        separation_factor = 3,

        force_unit_fuzzy_goto_behavior = false
      }
    },

    path_finder=
    {
      fwd2bwd_ratio = 5,
      goal_pressure_ratio = 2,
      max_steps_worked_per_tick = 1000,
      max_work_done_per_tick = 8000,
      use_path_cache = true,
      short_cache_size = 5,
      long_cache_size = 25,
      short_cache_min_cacheable_distance = 10,
      short_cache_min_algo_steps_to_cache = 50,
      long_cache_min_cacheable_distance = 30,
      cache_max_connect_to_cache_steps_multiplier = 100,
      cache_accept_path_start_distance_ratio = 0.2,
      cache_accept_path_end_distance_ratio = 0.15,
      negative_cache_accept_path_start_distance_ratio = 0.3,
      negative_cache_accept_path_end_distance_ratio = 0.3,
      cache_path_start_distance_rating_multiplier = 10,
      cache_path_end_distance_rating_multiplier = 20,
      stale_enemy_with_same_destination_collision_penalty = 30,
      ignore_moving_enemy_collision_distance = 5,
      enemy_with_different_destination_collision_penalty = 30,
      general_entity_collision_penalty = 10,
      general_entity_subsequent_collision_penalty = 3,
      extended_collision_penalty = 3,
      max_clients_to_accept_any_new_request = 10,
      max_clients_to_accept_short_new_request = 100,
      direct_distance_to_consider_short_request = 100,
      short_request_max_steps = 1000,
      short_request_ratio = 0.5,
      min_steps_to_check_path_find_termination = 2000,
      start_to_goal_cost_multiplier_to_terminate_path_find = 2000.0,
      overload_levels = {0, 100, 500},
      overload_multipliers = {2, 3, 4},
      negative_path_cache_delay_interval = 20
    },

    max_failed_behavior_count = 3,

    difficulty_settings =
    {
       recipe_difficulty = defines.difficulty_settings.recipe_difficulty.normal,
       technology_difficulty = defines.difficulty_settings.technology_difficulty.normal,
       -- technology_price_multiplier -- not needed
       -- research_queue_setting -- not needed
    }
  }
})
